### ISSUE TITLE
Conditional Hyperopting of Parameters - Possible?

### ISSUE THREAD

**Issue by seanmavley** [freqtrade/freqtrade#11034]
## Describe your environment

  * Operating system: Ubuntu 24.04 LTS WSL
  * Python Version: Python 3.10.12
  * CCXT version: ccxt==4.4.25
  * Freqtrade Version: freqtrade 2024.11
  
## Your question

*Ask the question you have not been able to find an answer in the [Documentation](https://www.freqtrade.io/en/latest/)*

Let's say I have this strategy

```
from freqtrade.strategy import IStrategy, IntParameter
import talib.abstract as ta
import pandas as pd
import freqtrade.vendor.qtpylib.indicators as qtpylib

class SimpleEMA(IStrategy):

    timeframe = '5m'
    startup_candle_count = 720
    stoploss = -0.99
    minimal_roi = {"0": 0.2}

    a_timeperiod = IntParameter(5, 18, default=10, space="buy")
    b_timeperiod = IntParameter(60, 80, default=10, space="buy")

    x_timeperiod = IntParameter(5, 18, default=10, space="buy")
    y_timeperiod = IntParameter(60, 80, default=10, space="buy")

    c_timeperiod = IntParameter(5, 18, default=10, space="buy")
    d_timeperiod = IntParameter(60, 80, default=10, space="buy")

    def leverage(self, pair: str, current_time, current_rate, proposed_leverage: float, **kwargs) -> float:
        return 8.0

    def populate_indicators(self, dataframe: pd.DataFrame, metadata: dict) -> pd.DataFrame:

        dataframe['a'] = ta.EMA(dataframe['ha_close'], timeperiod=self.a_timeperiod.value)
        dataframe['b'] = ta.EMA(dataframe['ha_close'], timeperiod=self.b_timeperiod.value)

        dataframe['x'] = ta.EMA(dataframe['ha_close'], timeperiod=self.x_timeperiod.value)
        dataframe['y'] = ta.EMA(dataframe['ha_close'], timeperiod=self.y_timeperiod.value)

        dataframe['c'] = ta.EMA(dataframe['ha_close'], timeperiod=self.c_timeperiod.value)
        dataframe['d'] = ta.EMA(dataframe['ha_close'], timeperiod=self.d_timeperiod.value)

        return dataframe

    def populate_entry_trend(self, dataframe, metadata):

        if (condition_1):
            long_condition = (
                qtpylib.crossed_above(dataframe['a'], dataframe['b'])
            )
        elif (condition_2):
            long_condition = (
                qtpylib.crossed_below(dataframe['x'], dataframe['y'])
            )
        else:
            long_condition = (
                qtpylib.crossed_above(dataframe['c'], dataframe['d'])
            )

        dataframe.loc[long_condition, 'enter_long'] = 1
        return dataframe

    def populate_exit_trend(self, dataframe, metadata):
        return dataframe
```

My questions:

- During hyperopting, will all the parameters be set, or only the matching condition values are tweaked on the fly?
- If not, is there a way to let hyperopt only spend resource on the matching entry conditions params only at any iteration in time? As in, if `condition_1` matches, THEN hyperopt tweaks the values of the corresponding `a_timeperiod` and `b_timeperiod`
- If none of the above is anywhere close to what actually happens with hyperopting, kindly enlighten me

My current understanding of hyperopting:

Hyperopting is an automated version of backtesting which swaps out the range of parameter values in millions (if not thousands) of combinations, and picks the best outputs in the end.

**Comment by xmatthias** [comment]
Your sample doesn't really make sense - as "condition1" is not defined ... but the answer is no.

All parameters will be set  - and (after an initial random space) modified depending on the loss function's outcome.
Now which parameter is modified is determined by the search algorithm - it's usually several (it's bayesian search in a hyper-dimensional space).

Now unfortunately, scikit-optimize doesn't support hierarchical spaces.
So if you have 2 conditions - say "use_bb" (true/false)  and "bb_value" (an actual range) - then hyperopt will still change the `bb_value` parameters - even if "use_bb" is false - so the value is not actually used (in fact, the algorithm has no way to know if you're ACTUALLY using a value.

It'll also not necessarily try all combinations (depending on the search space).

What you can however do is set a parameter to "optimize=False" - which means it'll remain static and will not be included in the optimization.

---

to simplify it - think about it as a valley surrounded by hills (you're looking for the deepest point).
You start somewhere halfway up the hill - then you try a few points - determine the direction to go in to reach the bottom of the valley.
You may also climb a pass (if it's not a high pass and you started near it) - and then descend in another valley, trying to see if this valley is deeper.

Now hyperopt starts this process with 30 random points - trying to search for the lowest/deepest valley (smallest loss).
This may mean that it starts at the best point (unlikely) - or it may also mean that it'll miss 20 other valleys that would be deeper.

Hyperopt does the same thing - just not in a 3d space (valley with hills) - but in a Ndimensional space - which essentially we can neither visualize nor grap - as we live in a 3d world - not in a 20 dimensional world.

**Comment by seanmavley** [comment]
> Your sample doesn't really make sense - as "condition1" is not defined ... but the answer is no.

Didn't include condition for brevity, but it could be as simple as 

```
condition1 = metadata['pair'] == "BTC/USDT:USDT"
```

> All parameters will be set - and (after an initial random space) modified depending on the loss function's outcome.
Now which parameter is modified is determined by the search algorithm - it's usually several (it's bayesian search in a hyper-dimensional space).

Right. Understood.

> It'll also not necessarily try all combinations (depending on the search space).

Yes, makes sense. I've since been reading the various loss functions to understand what's happening. Fascinating.

> to simplify it - think about it as a valley surrounded by hills (you're looking for the deepest point).
You start somewhere halfway up the hill - then you try a few points - determine the direction to go in to reach the bottom of the valley.
You may also climb a pass (if it's not a high pass and you started near it) - and then descend in another valley, trying to see if this valley is deeper.

> Now hyperopt starts this process with 30 random points - trying to search for the lowest/deepest valley (smallest loss).
This may mean that it starts at the best point (unlikely) - or it may also mean that it'll miss 20 other valleys that would be deeper.

> Hyperopt does the same thing - just not in a 3d space (valley with hills) - but in a Ndimensional space - which essentially we can neither visualize nor grap - as we live in a 3d world - not in a 20 dimensional world.

Yes, I understand. Thanks for the explanation
